rsocket-core/src/main/java/io/rsocket/util/ReconnectingRSocket.java

```diff
@@ -0,0 +1,539 @@
+import reactor.core.scheduler.Scheduler;
+import reactor.core.scheduler.Schedulers;
+import reactor.util.annotation.Nullable;
+import reactor.util.context.Context;
+
+@SuppressWarnings("unchecked")
+public class ReconnectingRSocket implements CoreSubscriber<RSocket>, RSocket, Runnable {
+
+  private final long backoffMinInMillis;
+  private final long backoffMaxInMillis;
+  private final Mono<? extends RSocket> source;
+  private final Scheduler scheduler;
+  private final Predicate<Throwable> errorPredicate;
+  private final MonoProcessor<Void> onDispose;
+
+  RSocket value;
+
+  volatile Consumer<RSocket>[] subscribers;
+
+  @SuppressWarnings("rawtypes")
+  static final AtomicReferenceFieldUpdater<ReconnectingRSocket, Consumer[]> SUBSCRIBERS =
+      AtomicReferenceFieldUpdater.newUpdater(
+          ReconnectingRSocket.class, Consumer[].class, "subscribers");
+
+  @SuppressWarnings("rawtypes")
+  static final Consumer<RSocket>[] EMPTY_UNSUBSCRIBED = new Consumer[0];
+
+  @SuppressWarnings("rawtypes")
+  static final Consumer<RSocket>[] EMPTY_SUBSCRIBED = new Consumer[0];
+
+  @SuppressWarnings("rawtypes")
+  static final Consumer<RSocket>[] TERMINATED = new Consumer[0];
+
+  static final ClosedChannelException ON_CLOSE_EXCEPTION = new ClosedChannelException();
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  ReconnectingRSocket(
+      Mono<? extends RSocket> source,
+      Predicate<Throwable> errorPredicate,
+      Scheduler scheduler,
+      long backoffMinInMillis,
+      long backoffMaxInMillis) {
+
+    this.source = source;
+    this.backoffMinInMillis = backoffMinInMillis;
+    this.backoffMaxInMillis = backoffMaxInMillis;
+    this.scheduler = scheduler;
+    this.errorPredicate = errorPredicate;
+    this.onDispose = MonoProcessor.create();
+
+    SUBSCRIBERS.lazySet(this, EMPTY_UNSUBSCRIBED);
+  }
+
+  public void subscribe(Consumer<RSocket> actual) {
+    if (!add(actual)) {
+      actual.accept(value);
+    }
+  }
+
+  @Override
+  public void onSubscribe(Subscription subscription) {
+    subscription.request(Long.MAX_VALUE);
+  }
+
+  @Override
+  public void onComplete() {
+    final RSocket value = this.value;
+
+    if (value == null) {
+      reconnect();
+    } else {
+      value.onClose().subscribe(null, null, () -> resubscribeWhen(ON_CLOSE_EXCEPTION));
+      Consumer<RSocket>[] array = SUBSCRIBERS.getAndSet(this, TERMINATED);
+      for (Consumer<? super RSocket> as : array) {
+        as.accept(value);
+      }
+    }
+  }
+
+  @Override
+  public void onError(Throwable t) {
+    reconnect();
+  }
+
+  @Override
+  @SuppressWarnings("unchecked")
+  public void onNext(@Nullable RSocket value) {
+    this.value = value;
+  }
+
+  @Override
+  public void run() {
+    source.subscribe(this);
+  }
+
+  private void reconnect() {
+    ThreadLocalRandom random = ThreadLocalRandom.current();
+    long nextRandomDelay = random.nextLong(backoffMinInMillis, backoffMaxInMillis);
```

Comment:
Could the backoff strategy be hidden behind a Flux<Long> without losing much?  Or would that negate "high-performant low-overhead"?